### PR TITLE
Fix bugs download all submissions

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
None

#### What's this PR do?
- Adds server side check for max size limit
- fix zip file creation code

#### How should this be manually tested?
- change storage to s3 and upload assignments and download all
- change storage to file system and upload assignments and download all
- use big files. MIT has max file upload size 20mb

#### Where should the reviewer start?
- Adds logs on all student submissions download
- add server side validation of max file size

### CODE REVIEW

File: edx_sga/tasks.py [hunk 1]
```diff
@@ -61,10 +61,15 @@ def _compress_student_submissions(zip_file_path, block_id, course_id, locator):
     student_submissions = _get_student_submissions(block_id, course_id, locator)
     if not student_submissions:
         return
+
+    log.info("Fetched all submission count: %d for zip path: %s ", len(student_submissions), zip_file_path)
     # Build the zip file in memory
     zip_file_bytes = BytesIO()
-    with zipfile.ZipFile(zip_file_bytes, 'w') as zip_pointer:
+    with zipfile.ZipFile(zip_file_bytes, 'w', compression=zipfile.ZIP_DEFLATED, allowZip64=True) as zip_pointer:
```
Comment (noisecapella) [inline edx_sga/tasks.py]: I don't think we want `allowZip64` here. I don't think we will need to handle zip files larger than 2gb after compression, and using this option makes the zip files incompatible with certain command line tools

Comment (noisecapella) [reply edx_sga/tasks.py]: Adding the `compression` flag is a very good idea though

File: edx_sga/sga.py [hunk 2]
```diff
@@ -146,6 +147,18 @@ class StaffGradedAssignmentXBlock(StudioEditableXBlockMixin, ShowAnswerXBlockMix
         help=_("When the annotated file was uploaded")
     )
 
+    @classmethod
+    def file_size_under_limit(cls, file_obj):
```
Comment (noisecapella) [inline edx_sga/sga.py]: It looks like we check this limit in the Javascript, and then again with the web server (it returns a 413 indicating the file is too large). I'm not sure we should add this check here.

Also this should be `file_size_over_limit`

Comment (amir-qayyum-khan) [reply edx_sga/sga.py]: just in case someone use api to submit file, extra safty

**Comment by amir-qayyum-khan** [comment]
Done @noisecapella

**Comment by gsidebo** [comment]
Works in my ginkgo vagrant VM. Looking at code now

**Comment by amir-qayyum-khan** [comment]
Awesome, please check it with big files as mention in description, because issue is there

File: edx_sga/sga.py [hunk 3]
```diff
@@ -221,17 +234,24 @@ def save_sga(self, data, suffix=''):
...
         Save a students submission file.
         """
         require(self.upload_allowed())
         user = self.get_real_user()
         require(user)
+        upload = request.params['assignment']
+        if self.file_size_over_limit(upload.file):
+            raise JsonHandlerError(
+                413, 'Unable to upload file. Max size limit is {size}'.format(
+                    size=self.STUDENT_FILEUPLOAD_MAX_SIZE
```
Comment (gsidebo) [inline edx_sga/sga.py]: `file_size_over_limit` prefers the Django settings value for `STUDENT_FILEUPLOAD_MAX_SIZE` over the `STUDENT_FILEUPLOAD_MAX_SIZE` class property. If that Django settings value is set, then you're returning the wrong upload limit here. I think it makes sense to rename the `STUDENT_FILEUPLOAD_MAX_SIZE` property to `DEFAULT_STUDENT_FILEUPLOAD_MAX_SIZE`, create a new `@property` called `student_upload_max_size` or something like that, and implement it like this:

```python
@property
def student_upload_max_size(self):
    return getattr(
        settings, 
        "STUDENT_FILEUPLOAD_MAX_SIZE",
        self.DEFAULT_STUDENT_FILEUPLOAD_MAX_SIZE
    )
```

Then `file_size_over_limit` and this handler method would both read `self.student_upload_max_size` to determine the effective max upload size.

File: edx_sga/sga.py [hunk 4]
```diff
@@ -449,27 +469,43 @@ def prepare_download_submissions(self, request, suffix=''):  # pylint: disable=u
...
+            log.info("Zip file already available for block: %s for staff: %s", location, user.username)
             assignments = self.get_sorted_submissions()
             if assignments:
                 last_assignment_date = assignments[0]['timestamp'].astimezone(pytz.utc)
+                log.info(
+                    "Last assignment upload time: %s for block: %s for staff: %s",
+                    last_assignment_date,
+                    location,
+                    user.username
+                )
```
Comment (gsidebo) [inline edx_sga/sga.py]: I think this log message can be removed. The message at line 492 gives the user the same information

File: edx_sga/sga.py [hunk 5]
```diff
@@ -449,27 +469,43 @@ def prepare_download_submissions(self, request, suffix=''):  # pylint: disable=u
         user = self.get_real_user()
         require(user)
         zip_file_ready = False
+        location = unicode(self.location)
 
         if self.is_zip_file_available(user):
+            log.info("Zip file already available for block: %s for staff: %s", location, user.username)
```
Comment (gsidebo) [inline edx_sga/sga.py]: ~~`for staff:`~~ --> `for user:` or `for instructor:`

File: edx_sga/sga.py [hunk 6]
```diff
@@ -449,27 +469,43 @@ def prepare_download_submissions(self, request, suffix=''):  # pylint: disable=u
...
+                )
                 zip_file_path = get_zip_file_path(
                     user.username,
                     self.block_course_id,
                     self.block_id,
                     self.location
                 )
                 zip_file_time = get_file_modified_time_utc(zip_file_path)
+                log.info(
+                    "Last assignment upload time: %s, last zip file time: %s for block: %s for staff: %s",
```
Comment (gsidebo) [inline edx_sga/sga.py]: 1. ~~`for staff:`~~ --> `for user:` or `for instructor:`
1. ~~`last zip file time:`~~ --> `zip file modified time:`

File: edx_sga/sga.py [hunk 7]
```diff
@@ -449,27 +469,43 @@ def prepare_download_submissions(self, request, suffix=''):  # pylint: disable=u
...
+                    zip_file_time,
+                    location,
+                    user.username
+                )
                 # if last zip file is older the last submission then recreate task
                 if zip_file_time >= last_assignment_date:
                     zip_file_ready = True
 
         if not zip_file_ready:
+            log.info("Creating new zip file for block: %s for staff: %s", location, user.username)
```
Comment (gsidebo) [inline edx_sga/sga.py]: ~~`for staff:`~~ --> `for user:` or `for instructor:`

File: edx_sga/tasks.py [hunk 8]
```diff
@@ -90,8 +96,11 @@ def zip_student_submissions(course_id, block_id, locator_unicode, username):
         username (unicode): user name of the staff user requesting the zip file
     """
     locator = BlockUsageLocator.from_string(locator_unicode)
+    log.info("creating zip file for course locator: %s", locator)
     zip_file_path = get_zip_file_path(username, course_id, block_id, locator)
+    log.info("Zip file relative path: %s", zip_file_path)
     if default_storage.exists(zip_file_path):
+        log.info("Delete zip file if exist at path: %s", zip_file_path)
```
Comment (gsidebo) [inline edx_sga/tasks.py]: `log.info("Deleting already-existing zip file at path: %s", zip_file_path)`

File: edx_sga/tasks.py [hunk 9]
```diff
@@ -90,8 +96,11 @@ def zip_student_submissions(course_id, block_id, locator_unicode, username):
         username (unicode): user name of the staff user requesting the zip file
     """
     locator = BlockUsageLocator.from_string(locator_unicode)
+    log.info("creating zip file for course locator: %s", locator)
     zip_file_path = get_zip_file_path(username, course_id, block_id, locator)
+    log.info("Zip file relative path: %s", zip_file_path)
```
Comment (gsidebo) [inline edx_sga/tasks.py]: Please combine this message with the one at line 99 into a single message:

```python
log.info("Creating zip file for course: %s at path: %s" ,locator, zip_file_path)
```

File: edx_sga/tasks.py [hunk 10]
```diff
@@ -61,21 +59,29 @@ def _compress_student_submissions(zip_file_path, block_id, course_id, locator):
...
-                    os.path.basename(submission_file_path)
+
+    log.info("Fetched all submission count: %d for zip path: %s ", len(student_submissions), zip_file_path)
+    # Build the zip file in memory using temporary file wrapper  specialized to switch
+    # from StringIO to a real file when it exceeds a certain size or when a fileno is needed.
+    with tempfile.SpooledTemporaryFile() as tmp:
+        with zipfile.ZipFile(tmp, 'w', compression=zipfile.ZIP_DEFLATED) as zip_pointer:
+            for student_username, submission_file_path in student_submissions:
+                log.info(
+                    "Zipping file in memory student: %s, path: %s ", student_username, submission_file_path
```
Comment (gsidebo) [inline edx_sga/tasks.py]: `"Creating zip file for student: %s, submission path: %s ", student_username, submission_file_path`

File: edx_sga/tasks.py [hunk 11]
```diff
@@ -61,21 +59,29 @@ def _compress_student_submissions(zip_file_path, block_id, course_id, locator):
...
-    zip_file_bytes = BytesIO()
-    with zipfile.ZipFile(zip_file_bytes, 'w') as zip_pointer:
-        for student_username, submission_file_path in student_submissions:
-            with default_storage.open(submission_file_path, 'rb') as destination_file:
-                filename_in_zip = '{}_{}'.format(
-                    student_username,
-                    os.path.basename(submission_file_path)
+
+    log.info("Fetched all submission count: %d for zip path: %s ", len(student_submissions), zip_file_path)
+    # Build the zip file in memory using temporary file wrapper  specialized to switch
```
Comment (gsidebo) [inline edx_sga/tasks.py]: Nitpick: double-space before 'specialized'

File: edx_sga/tasks.py [hunk 12]
```diff
@@ -61,21 +59,29 @@ def _compress_student_submissions(zip_file_path, block_id, course_id, locator):
...
-    # Build the zip file in memory
-    zip_file_bytes = BytesIO()
-    with zipfile.ZipFile(zip_file_bytes, 'w') as zip_pointer:
-        for student_username, submission_file_path in student_submissions:
-            with default_storage.open(submission_file_path, 'rb') as destination_file:
-                filename_in_zip = '{}_{}'.format(
-                    student_username,
-                    os.path.basename(submission_file_path)
+
+    log.info("Fetched all submission count: %d for zip path: %s ", len(student_submissions), zip_file_path)
```
Comment (gsidebo) [inline edx_sga/tasks.py]: `log.info("Compressing %d student submissions to path: %s ", len(student_submissions), zip_file_path)`

**Comment by amir-qayyum-khan** [comment]
Done @gsidebo

File: edx_sga/tasks.py [hunk 13]
```diff
@@ -61,21 +59,31 @@ def _compress_student_submissions(zip_file_path, block_id, course_id, locator):
...
+                    filename_in_zip = '{}_{}'.format(
+                        student_username,
+                        os.path.basename(submission_file_path)
+                    )
+                    zip_pointer.writestr(filename_in_zip, destination_file.read())
+        # Reset file pointer
+        tmp.seek(0)
+        # Write the bytes of the in-memory zip file to an actual file
+        log.info(
+            "Moving zip file from memory to storage for: zip path: %s ", zip_file_path
```
Comment (gsidebo) [inline edx_sga/tasks.py]: `"Moving zip file from memory to storage at path: %s ", zip_file_path`

**Comment by amir-qayyum-khan** [comment]
@mitodl/devops @pdpinch  done

**Pull request merged**